cardano-cli/src/Cardano/CLI/Slot.hs

```diff
@@ -0,0 +1,329 @@
+
+-- | The current system time, compared to the blockchain start time of the given
+-- 'TimeInterpreter'.
+--
+-- If the current time is before the system start (this would only happen when
+-- launching testnets), the relative time is reported as 0.
+currentRelativeTime :: MonadIO m => TimeInterpreter n -> m RelativeTime
+currentRelativeTime = liftIO . getCurrentTimeRelativeFromStart . blockchainStartTime
+
+-- | Takes a motivation of why @TimeInterpreter@ shouldn't fail interpreting
+-- queries.
+--
+-- Unexpected @PastHorizonException@s will be thrown in IO, and traced with
+-- Error severity along with the provided motivation.
+neverFails
+  :: TimeInterpreter (ExceptT PastHorizonException IO)
+  -> TimeInterpreter IO
+neverFails = f . hoistTimeInterpreter (runExceptT >=> eitherToIO)
+  where
+    eitherToIO (Right x) = pure x
+    eitherToIO (Left e) = throwIO e
+
+    f (TimeInterpreter getI ss) = TimeInterpreter
+      { interpreter = getI
+      , blockchainStartTime = ss
+      }
+
+-- | Change the underlying monad of the TimeInterpreter with a natural
+-- transformation.
+hoistTimeInterpreter
+  :: (forall a. m a -> n a)
+  -> TimeInterpreter m
+  -> TimeInterpreter n
+hoistTimeInterpreter f (TimeInterpreter getI ss) = TimeInterpreter
+  { interpreter = f getI
+    -- NOTE: interpreter ti cannot throw PastHorizonException, but
+    -- this way we don't have to carry around yet another type parameter.
+  , blockchainStartTime = ss
+  }
+
+-- | The current system time, compared to the given blockchain start time.
+--
+-- If the current time is before the system start (this would only happen when
+-- launching testnets), let's just say we're in epoch 0.
+--
+-- TODO: Use io-sim-classes for easier testing.
+getCurrentTimeRelativeFromStart :: StartTime -> IO RelativeTime
+getCurrentTimeRelativeFromStart start = toRelativeTimeOrZero start <$> getCurrentTime
+
+-- | Convert an absolute time to a relative time. If the absolute time is before
+-- the system start, consider the relative time to be the system start
+-- time. This function can never fail.
+toRelativeTimeOrZero :: StartTime -> UTCTime -> RelativeTime
+toRelativeTimeOrZero start = fromMaybe (RelativeTime 0) . toRelativeTime start
+
+-- | Same as 'Cardano.toRelativeTime', but has error handling for times before
+-- the system start. No other functions in this module will accept UTC times.
+toRelativeTime :: StartTime -> UTCTime -> Maybe RelativeTime
+toRelativeTime (StartTime start) utc
+  | utc < start = Nothing
+  | otherwise = Just $ Cardano.toRelativeTime (SystemStart start) utc
+
+-- | Set up a 'TimeInterpreter' for a given start time, and an 'Interpreter'
+-- queried from the ledger layer.
+mkTimeInterpreter
+    :: Monad m
+    => StartTime
+    -> EpochSize
+    -> Cardano.SlotLength
+    -> TimeInterpreter (ExceptT PastHorizonException m)
+mkTimeInterpreter start sz len = TimeInterpreter
+    { interpreter = pure (mkInterpreter (neverForksSummary sz len))
```

Comment:
If I understand the context correctly, then the use of `neverForksSummary` is not the right way to acquire an `Interpreter`. IE We're concerned about syncing the Cardano chain, and that should does fork, so the "never forks" constructor for `Summary` doesn't fit this use-case.

I'll write my in-a-vacuum expectations in another comment.